modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/DoubleGaugeAdapter.java

```diff
@@ -10,33 +10,45 @@
 
 import io.opentelemetry.api.metrics.Meter;
 
+import java.util.Collections;
 import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * DoubleGaugeAdapter wraps an otel ObservableDoubleMeasurement
  */
-public class DoubleGaugeAdapter extends AbstractInstrument<io.opentelemetry.api.metrics.ObservableDoubleMeasurement>
+public class DoubleGaugeAdapter extends AbstractInstrument<io.opentelemetry.api.metrics.ObservableDoubleGauge>
     implements
         org.elasticsearch.telemetry.metric.DoubleGauge {
 
+    private final AtomicReference<ValueWithAttributes<Double>> valueWithAttributes;
+
     public DoubleGaugeAdapter(Meter meter, String name, String description, String unit) {
         super(meter, name, description, unit);
+        this.valueWithAttributes = new AtomicReference<>(new ValueWithAttributes(0.0, Collections.emptyMap()));
     }
 
     @Override
-    io.opentelemetry.api.metrics.ObservableDoubleMeasurement buildInstrument(Meter meter) {
+    io.opentelemetry.api.metrics.ObservableDoubleGauge buildInstrument(Meter meter) {
         var builder = Objects.requireNonNull(meter).gaugeBuilder(getName());
-        return builder.setDescription(getDescription()).setUnit(getUnit()).buildObserver();
+        return builder.setDescription(getDescription())
+            .setUnit(getUnit())
+            .buildWithCallback(
+                measurement -> measurement.record(
+                    valueWithAttributes.get().value(),
```

Comment:
This has to be stored in a local variable to avoid a race between the two calls to `valueWithAttributes.get()`.